lib/LedgerSMB/PGTimestamp.pm

```diff
@@ -0,0 +1,223 @@
+defers object creation to the superclass.
+
+=cut
+
+sub new {
+    my $class = shift;
+    my @args = @_;
+
+    if (! @args) {
+        my $self = {};
+        bless $self, $class;
+
+        $self->is_date(0);
+        $self->is_time(0);
+
+        return $self;
+    }
+    return $class->SUPER::new(@args);
+}
+
+
+=item add_interval(string $interval, optional integer $n)
+
+This adds $n * $interval to the date, defaulting to 1 if $n is not supplied.
+
+=cut
+
+sub add_interval {
+    my ($self,$interval,$n) = @_;
+
+    my %delta_names = (
+        day => 'days',
+        week => 'weeks',
+        month => 'months',
+        quarter => 'months',
+        year => 'years',
+    );
+    my $delta_name = $delta_names{$interval};
+    #Validate asked interval
+    die "Bad interval: $interval" if not defined $delta_name;
+
+    $n //= 1;    # Default to 1
+    $n *= MONTHS_PER_QUARTER if $interval eq 'quarter'; # A quarter is 3 months
+
+    my $has_time = $self->is_time();
+    $self->add($delta_name => $n, end_of_month => 'preserve');
+    $self->is_time($has_time);  # Make sure that is_time sticks
+
+    return $self;
+}
+
+=item from_input($date_string, [$format])
+
+Parses this from an input string according to the user's dateformat,
+unless C<$format> has been specified to override it.
+
+=cut
+
+our $formats = {
+    'YYYY-MM-DD' => ['%F'],
+    'DD-MM-YYYY' => ['%d-%m-%Y', '%d-%m-%y'],
+    'DD.MM.YYYY' => ['%d.%m.%Y', '%d.%m.%y'],
+    'DD/MM/YYYY' => ['%d/%m/%Y', '%D'],
+    'MM-DD-YYYY' => ['%m-%d-%Y', '%m-%d-%y'],
+    'MM/DD/YYYY' => ['%m/%d/%Y', '%m/%d/%y'],
+    'MM.DD.YYYY' => ['%m.%d.%Y', '%m.%d.%y'],
+      'YYYYMMDD' => ['%Y%m%d'],
+        'YYMMDD' => ['%y%m%d'],
+      'DDMMYYYY' => ['%d%m%Y'],
+        'DDMMYY' => ['%d%m%y'],
+      'MMDDYYYY' => ['%m%d%Y'],
+        'MMDDYY' => ['%m%d%y'],
+     'DDmonYYYY' => ['%d%b%Y', '%d%b%y']
+};
+
+sub from_input{
+    my ($self, $input) = @_;
+
+    local $@ = undef;
+    return $input if eval {$input->isa(__PACKAGE__)} && $input->is_date;
+
+    return __PACKAGE__->new()
+        if ! $input; # matches undefined as well as ''
+
+    my $dt = eval { strptime('%F %T', $input) }
+        // strptime('%F', $input);
+
+    die "Bad timestamp ($input)" if $input && ! $dt;
+    bless $dt, __PACKAGE__;
+    $dt->is_date(1);
+    $dt->is_time(($input && $input =~ /\:/) ? 1 : 0); # Define time
+    return $dt;
+}
+
+
+=item to_output(optional string $format)
+
+This returns the human readable formatted date.  If $format is supplied, it is
+used.  If $format is not supplied, the dateformat of the user is used.
+
+=cut
+
+sub to_output {
+    my ($self) = @_;
+    return '' if not $self->is_date();
+    my $fmt;
+    if (defined $LedgerSMB::App_State::User->{dateformat}){
+        $fmt = $LedgerSMB::App_State::User->{dateformat};
+    } else {
+        $fmt = '%F';
+    }
+    $fmt = $formats->{uc($fmt)}->[0] if defined $formats->{uc($fmt)};
```

Comment:
$formats->{$fmt} gives a list of possible patterns, but we're only trying the first of them here. Should we be testing all of the options?

For example 'DD-MM-YYYY' maps to  ['%d-%m-%Y', '%d-%m-%y'], which I tests for both 4 and 2 digit years. Albeit that is rather surprising behaviour and I'm not sure that we should be supporting 2 digit years as input.

Reply:
Completely agreed. I think 2-digit years are too confusing. I'll PR for PGDate seperately, along the same lines.